typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/helpers/graph.ts

```diff
@@ -0,0 +1,334 @@
+		const leaf = this.tree.leaves[name];
+		if (!leaf) {
+			throw new Error(`Leaf "${name}" not found`);
+		}
+
+		return leaf;
+	}
+
+	/**
+	 * Height of the graph
+	 *
+	 * Retrieve from here so you don't have to store a variable
+	 */
+	public getHeight() {
+		return this.height;
+	}
+
+	/**
+	 * Width of the graph
+	 *
+	 * Retrieve from here so you don't have to store a variable
+	 */
+	public getWidth() {
+		return this.width;
+	}
+
+	/**
+	 * Node spacing
+	 *
+	 * Retrieve from here so you don't have to store a variable
+	 */
+	public getNodeSpacing() {
+		return this.nodeSpacing;
+	}
+
+	/**
+	 * Check if a node is already at the given position
+	 */
+	private hasOverlap(x: number, y: number) {
+		for (const leaf of Object.values(this.tree.leaves)) {
+			if (Math.round(leaf.node.x) === Math.round(x) && Math.round(leaf.node.y) === Math.round(y)) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Builds the tree
+	 *
+	 * Calculates the x and y positions of each leaf
+	 */
+	private buildTree(
+		origin: GraphNode,
+		visited: Set<GraphNode> = new Set<GraphNode>(),
+		x: number = this.nodeSpacing,
+		y: number = this.height / 2
+	) {
+		while (this.hasOverlap(x, y)) {
+			y += this.nodeSpacing;
+		}
+
+		origin.x = x;
+		origin.y = y;
+
+		let positions = {
+			x: x + this.nodeSpacing,
+			y: y
+		};
+
+		if (origin.targets.length > 1) {
+			positions.y -= this.nodeSpacing * (origin.targets.length - 1) / 2;
+		}
+
+		origin.targets.forEach(targetName => {
+			const target = this.getNode(targetName);
+
+			if (!this.tree.leaves[target.name]) {
+				this.tree.leaves[target.name] = {
+					node: target,
+					leaves: []
+				};
+			}
+
+			const targetLeaf = this.getLeaf(target.name);
+			this.getLeaf(origin.name).leaves.push(targetLeaf);
+
+			// LPS (`Loop Prevention System`)
+			if (visited.has(target)) return;
+			visited.add(target);
+
+			this.buildTree(targetLeaf.node, visited, positions.x, positions.y);
+			positions.y += this.nodeSpacing;
+		});
+	}
+
+	/**
+	 * Calculates all the positions of the tree leaves
+	 *
+	 * If this method is called multiple times, it will only calculate the positions once
+	 */
+	private calculatePositions() {
+		if (this.calculated) return;
+
+		for (const node of this.nodes) {
+			this.tree.leaves[node.name] = {
+				node,
+				leaves: []
+			};
+		}
+
+		const visited = new Set<GraphNode>();
+		const spacedOutY = this.height / (this.tree.startNodes.length + 1);
+		const positions = { x: this.nodeSpacing, y: spacedOutY };
+
+		for (const name of this.tree.startNodes) {
+			const startNode = this.getNode(name);
+			this.buildTree(startNode, visited, positions.x, positions.y);
+
+			positions.y += spacedOutY;
+		}
+
+		this.calculated = true;
+
+		for (const leaf of Object.values(this.tree.leaves)) {
+			if (this.tree.startNodes.includes(leaf.node.name)) {
+				continue;
+			}
+
+			if (leaf.node.x === undefined || leaf.node.y === undefined) {
+
+				console.log(
```

Comment:
Extract to a variable, e.g:
```
const nodes = Object.values(this.tree.leaves).map(l => ({ node: l.node, leaves: l.leaves.map(le => le.node.name) }));
console.log(JSON.stringify(nodes, null, 2));
```
Wondering whether it wouldn't be better to include the nodes in the error instead:
```
throw new Error(`Node "${leaf.node.name}" has no position and is possibly isolated. Check the WF element targets:\n`
+ JSON.stringify(nodes, null, 2));
```

Reply:
Error gets drowned by the nodes, that's why it's separate